plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -145,32 +162,37 @@ public void invalidate(Partition partition)
         delegate.invalidate(partition);
     }
 
-    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, RemoteIterator<LocatedFileStatus> iterator, Path path)
+    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, Path path)
     {
         return new RemoteIterator<>()
         {
-            private final List<LocatedFileStatus> files = new ArrayList<>();
+            private int fileIndex;
 
             @Override
             public boolean hasNext()
                     throws IOException
             {
-                boolean hasNext = iterator.hasNext();
-                if (!hasNext) {
+                try {
+                    boolean hasNext = cachedValueHolder.hasNext(fileIndex);
+                    // Update cache weight of cachedValueHolder for a given path.
                     // The cachedValueHolder acts as an invalidation guard. If a cache invalidation happens while this iterator goes over
                     // the files from the specified path, the eventually outdated file listing will not be added anymore to the cache.
-                    cache.asMap().replace(path, cachedValueHolder, new ValueHolder(files));
+                    cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);
```

Comment:
what does it do?

Reply:
I'm assuming that this is a way to notify the weigher that the size of the cached item has changed.
@sopel39 why not updating the cache only once (when all the files have been retrieved) ?

Reply:
`.replace(k, v, v)` can be a no-op. It's not guaranteed that it notifies the weighter (even if it does so today)

btw i initially thought it's just a left-over, not something done intentionally.

Reply:
> also, there is a test for that

where is it?

Reply:
> where is it?

`TestCachingDirectoryLister.java#testConcurrentDirectoryListing`
(https://github.com/trinodb/trino/pull/11650/files#diff-c3ef0a7a18e7c6d314620a3ef5ab3cbea00e3ce0188ab37d962488c29929cc49R373)

Reply:
> line 411 (https://github.com/trinodb/trino/pull/11650/files#diff-c3ef0a7a18e7c6d314620a3ef5ab3cbea00e3ce0188ab37d962488c29929cc49R411)

This asserts on `CDL.getListCount`, so our internal accounting. It doesn't test the weighting (at least not directly).

when i remove `cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);`), the `TestCachingDirectoryLister` indeed fails at `assertThat(countingLister.getListCount()).isEqualTo(3);`
i guess this is because of the weight-triggered eviction, right?
if this is what's happening here, can you perhaps explicitly test the eviction, i.e. inspect the keys live in the cache?